python/pylibraft/pylibraft/neighbors/ivf_pq.pyx

```diff
@@ -0,0 +1,543 @@
+              const uint64_t* new_indices,       
+              uint64_t n_rows) #except +
+
+    cdef index[uint64_t] extend(const handle_t& handle,        
+              const index[uint64_t]& orig_index, 
+              const uint8_t* new_vectors,          
+              const uint64_t* new_indices,       
+              uint64_t n_rows) #except +
+
+    cdef void search(const handle_t& handle,
+                   const search_params& params,
+                   const index[uint64_t]& index,
+                   const float* queries,
+                   uint32_t n_queries,
+                   uint32_t k,
+                   uint64_t* neighbors,
+                   float* distances,
+                   device_memory_resource* mr) except +
+
+    cdef void search(const handle_t& handle,
+                   const search_params& params,
+                   const index[uint64_t]& index,
+                   const int8_t* queries,
+                   uint32_t n_queries,
+                   uint32_t k,
+                   uint64_t* neighbors,
+                   float* distances,
+                   device_memory_resource* mr) except +
+
+    cdef void search(const handle_t& handle,
+                   const search_params& params,
+                   const index[uint64_t]& index,
+                   const uint8_t* queries,
+                   uint32_t n_queries,
+                   uint32_t k,
+                   uint64_t* neighbors,
+                   float* distances,
+                   device_memory_resource* mr) except +
+
+
+def is_c_cont(cai):
+    dt = np.dtype(cai["typestr"])
+    return "strides" not in cai or \
+        cai["strides"] is None or \
+        cai["strides"][1] == dt.itemsize
+
+
+def _get_metric(metric):
+    SUPPORTED_DISTANCES = {
+        "l2_expanded": DistanceType.L2Expanded,
+        # TODO(tfeher): fix inconsistency: index building for L2SqrtExpanded is only supported by build, not by search.
+        # "euclidean": DistanceType.L2SqrtExpanded
+        "inner_product": DistanceType.InnerProduct
+    }
+    if metric not in SUPPORTED_DISTANCES:
+        raise ValueError("metric %s is not supported" % metric)
+    return SUPPORTED_DISTANCES[metric]
+
+
+def _check_input_array(cai, exp_dt, exp_rows=None, exp_cols=None):
+        if cai["typestr"] not in exp_dt:
+            raise TypeError("dtype %s not supported" % cai["typestr"])
+
+        if not is_c_cont(cai):
+            raise ValueError("Row major input is expected")
+
+        if exp_cols is not None and cai["shape"][1] != exp_cols:
+            raise ValueError("Incorrect number of columns, expected {} got {}" \
+                                 .format(exp_cols, cai["shape"][1]))
+
+        if exp_rows is not None and cai["shape"][0] != exp_rows:
+            raise ValueError("Incorrect number of rows, expected {} , got {}" \
+                                .format(exp_rows, cai["shape"][0]))
+        
+
+class IvfPq:
```

Comment:
Overall I'm really excited about this new API!

pylibraft is supposed to be very lightweight wrappers around the c++ layer, though, and as a result? I'd prefer if instead of following cuml/scikit-learn's "estimator" design, we design this so it directly corresponds to RAFT's C++ API.

This would make the ivfpq index itself a class with flat python functions in the pylibraft.neighbors.ivfpq package to build(), extend(), and search() the index. 

I think this also makes a more clear separation between cuml's python API and pylibraft, since cuml can continue to follow the "estimator" design style and not directly mirror pylibraft's API.

Reply:
That is a good point. Wrapping the index in its own class might help resolve the issue I am having with the exception handling. I will update the PR accordingly.

Reply:
Refactored, please have a look.

Reply:
Looks great! I'll give it a more thorough review tomorrow